### DISCUSSION TITLE
Caddy path

### DISCUSSION THREAD

**Discussion by Scialla** [indomus/forum#322] [GENERALE - Discussioni generiche]
Allora, visto che di la mi pigliate per il culo, lo scrivo pure di qua, così vi divertite anche qui :D 
Configurazione di caddy standard:
```
(https_header) {
  header {
    Strict-Transport-Security "max-age=31536000; includeSubdomains"
    X-XSS-Protection "1; mode=block"
    X-Content-Type-Options "nosniff"
    X-Frame-Options "SAMEORIGIN"
    Referrer-Policy "same-origin"
  }
https://pippo.dominio.net {
  import https_header
  reverse_proxy http://127.0.0.1:8123
}
```
e fin qui, 443 aperta, tutto funziona e HA si apre...
```
https://pippo.dominio.net/ha {
  import https_header
  reverse_proxy http://127.0.0.1:8123
}
```
questo invece non funziona
```
https://ha.pippo.dominio.net {
  import https_header
  reverse_proxy http://127.0.0.1:8123
}
```
e neppure questo...
non trovo esempi online, dove sbaglio?

C'è qualcuno che mi vuole bene? :) :) :)
Che poi non che mi sia indispensabile eh, ma non capisco e mi rode un pò...

**Comment by Newreader** [comment]
Non so cosa o meglio dove sia "di là", comunque provo a scriverti la mia considerazione.
Secondo me (come riportato anche dalla guida) dopo https://    andrebbe specificata la porta dove vengono indirizzate le richieste https (nella guida  9000) che successivamente vengono reindirizzate verso l'indirizzo interno che hai specificato.

**Reply by Scialla** [reply]
con la porta già va, così come senza, ma è il path che non va, ma ne abbiamo parlato, mi sa che proprio non si può fare purtroppo...

cmq grazie

**Reply by snakuzzo** [reply]
Il discorso sul path dipende soprattutto da com'è scritta l'applicazione che stai proxando

**Reply by Scialla** [reply]
Si ma non va neanche HA, non va la pagina del modem, ne altro. Restituisce sempre 404

**Reply by Newreader** [reply]
provo ancora una volta a scrivere le mie considerazioni.
Nella sezione http: nel file configuration.yaml, hai provato ad inserire, tra le altre, la linea:
base_url: https://tuodominio
con tuodominio che va ovviamente sostituito con quello di tuo interesse?

**Reply by snakuzzo** [reply]
`base_url` è deprecato...anzi non puoi proprio più usarlo
https://www.home-assistant.io/integrations/http/

**Reply by Newreader** [reply]
È vero! Dalla versione 0.110 di HA base_url è stato sostituito da external_url (basato su FQDN) e internal_url (basato su IP locale). Chiedo venia, al momento di scrivere la mia considerazione mi è sfuggito, anche perché fuorviato da alcune guide e discussioni lette sulla community.home-assistant.io.
Ho letto nei commenti precedenti che se ne è già parlato. Potete indicarmi cortesemente dove, in modo che io possa leggere tutte le considerazioni fatte. Questo, ovviamente, perché è un discorso che mi interessa approfondire.

**Reply by Scialla** [reply]
Ne abbiamo parlato privatamente non sul forum. Purtroppo è un casino, lascio perdere..

**Reply by snakuzzo** [reply]
> È vero! Dalla versione 0.110 di HA base_url è stato sostituito da external_url (basato su FQDN) e internal_url (basato su IP locale). Chiedo venia, al momento di scrivere la mia considerazione mi è sfuggito, anche perché fuorviato da alcune guide e discussioni lette sulla community.home-assistant.io.
> Ho letto nei commenti precedenti che se ne è già parlato. Potete indicarmi cortesemente dove, in modo che io possa leggere tutte le considerazioni fatte. Questo, ovviamente, perché è un discorso che mi interessa approfondire.

Sì ma nessuno dei tre c'entra/entrava niente con la raggiungibilità dall'esterno di HA. E qui è spiegato tutto...mi ha risposto sommo Frenck mesi e mesi fa...

![image](https://user-images.githubusercontent.com/7205371/119197458-0bba2800-ba88-11eb-9f14-78044f33031b.png)

**Comment by alabruna** [comment]
Avevo provato anch'io a usare i path (https://pippo.dominio.net/ha) ma come giustamente scritto sopra il più delle volte non vanno a causa di come sono scritte le applicazioni.
Detto ciò la versione con il sottodominio (https://ha.pippo.dominio.net) dovrebbe andare, ma bisogna fare attenzione all'ordine in cui si mettono le direttive sul Caddyfile, ovvero vanno definiti prima tutti i sottodomini e poi il dominio principale.

(https_header) {
  header {
    Strict-Transport-Security "max-age=31536000; includeSubdomains"
    X-XSS-Protection "1; mode=block"
    X-Content-Type-Options "nosniff"
    X-Frame-Options "SAMEORIGIN"
    Referrer-Policy "same-origin"
}
https://ha.pippo.dominio.net {
  import https_header
  reverse_proxy http://127.0.0.1:8123
}
https://pippo.dominio.net {
  import https_header
  reverse_proxy http://127.0.0.1:8123
}

Ho fatto la prova con duckdns e così mi funziona.

**Reply by Scialla** [reply]
Grazie, stasera ci provo, mi ero preso una settimana di vacanza.... faccio sapere :)

**Reply by Scialla** [reply]
niente, era no-ip che rompeva le balle, non reindirizza automaticamente i sottodomini di 4 livello, mentre duckdns lo fa.....
Funziona, grazie

**Comment by 2Max2Pax** [comment]
Ciao, ho qualche problema con Caddy.
Oltre alla 443 ho aperto le porte dalla 9000 alla 9005 per fare delle prove.
HA funziona correttamente nella rete interna (http://ip.del.server.ha:7000) (porta 7000 che ho arbitrariamente scelto per fare questo test).

Questo e' il **Caddyfile** :


```
(https_header) {
  header {
    Strict-Transport-Security "max-age=31536000; includeSubdomains"
    X-XSS-Protection "1; mode=block"
    X-Content-Type-Options "nosniff"
    X-Frame-Options "SAMEORIGIN"
    Referrer-Policy "same-origin"
  }
}


#questo funziona:
https://pr.pippo.dominio.net {
  respond "Prova 2"
  file_server
}

#questo non funziona:
https://ha.pippo.dominio.net:9000 {
  import https_header
  reverse_proxy http://ip.del.server.ha:7000
}

#questo non funziona:
https://hh.pippo.dominio.net {
  import https_header
  reverse_proxy http://127.0.0.1:7000
}

#questo non funziona:
https://pippo.dominio.net:9001 {
  respond "Prova 2"
  file_server
}

#questo non funziona:
https://pippo.dominio.net:9002 {
  import https_header
  reverse_proxy https://ip.del.server.ha:7000
}

#questo funziona:
https://pippo.dominio.net {
  respond "Prova 1"
  file_server
}
```

Se apro la porta 7000 sul router allora [https://pippo.dominio.net:7000](url) si collega correttamente ad HA, ho testato anche le due porte che uso per SSH senza successo.
Tutti i test li ho fatti con il cellulare, 
Non riesco a capire dove sbaglio.

**Reply by snakuzzo** [reply]
Ciao. Non basta dire "non funziona". Dovresti dettagliare meglio la cosa indicando che problemi hai.
Timeout? codice di errore HTTP che ricevi lato client ? Errori sul log di Caddy ?
E poi... crea un file di log specifico per ogni servizio che hai configurato

Es.

```
https://pippo.dominio.net:9002 {
	import https_header
	reverse_proxy http:127.0.0.1:7000
	log {
		output file /var/log/caddy/pippo.log {
			roll_size 5MiB
			roll_keep 2
			roll_keep_for 48h
		}
	}
}
```

Ad ogni modo, in base all'ultima cosa che hai scritto sull'apertura della porta 7000 e sul funzionamento di HA, qualcosa non torna. Lato Caddy non vedo alcun servizio configurato sulla porta 7000 e se ti risponde HA vuol dire quindi che da Caddy nemmeno ci passi

**Comment by 2Max2Pax** [comment]
Ho fatto altre prove anche con i log.


```
https://ha1.pippo.dominio.net { 
  import https_header
  reverse_proxy http://192.168.1.67:7000
  log {
		output file /var/log/caddy/ha10.log {
			roll_size 5MiB
			roll_keep 2
			roll_keep_for 48h
		}
	}
}
```
Safari risponde **400: Bad request** , log:
```
{
"level":"error",
"ts":1628669095.5305185,
"logger":"http.log.access.log0",
"msg":"handled request",
"request":{
	"remote_addr":"IP.DEL.MIO.CELLULARE:54899",
	"proto":"HTTP/2.0",
	"method":"GET",
	"host":"ha1.pippo.dominio.net",
	"uri":"/",
	"headers":{
			"Accept-Encoding":["gzip, deflate, br"],
			"User-Agent":["Mozilla/5.0 (iPhone; CPU iPhone OS 14_7_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.1.2 Mobile/15E148 Safari/604.1"],
			"Accept":["text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8"],
			"Accept-Language":["it-it"]
			},
	"tls":{
		"resumed":false,
		"version":772,
		"cipher_suite":4867,
		"proto":"h2",
		"proto_mutual":true,
		"server_name":"ha1.pippo.dominio.net"
		 }
	},
	"common_log":"IP.DEL.MIO.CELLULARE - - [11/Aug/2021:08:04:55 +0000] \"GET / HTTP/2.0\" 400 16",
	"duration":0.004476719,
	"size":16,
	"status":400,
	"resp_headers":{
		"Server":["Caddy","Python/3.9 aiohttp/3.7.4.post0"],
		"Referrer-Policy":["same-origin"],
		"Strict-Transport-Security":["max-age=31536000; includeSubdomains"],
		"X-Frame-Options":["SAMEORIGIN"],
		"Content-Length":["16"],
		"X-Content-Type-Options":["nosniff"],
		"X-Xss-Protection":["1; mode=block"],
		"Date":["Wed, 11 Aug 2021 08:04:55 GMT"],
		"Content-Type":["text/plain; charset=utf-8"]
		}
}
```

Altro test (ha2):

```
https://ha2.pippo.dominio.net {
  import https_header
  reverse_proxy http://127.0.0.1:7000
  log {
		output file /var/log/caddy/ha20.log {
			roll_size 5MiB
			roll_keep 2
			roll_keep_for 48h
		}
	}
}
```
In questo caso safari mi chiede **Vuoi scaricare "ha2.pippo.dominio.net.dms"?** (notare l'aggiunta di **dms** alla fine dell'indirizzo.
e questo e' il log:
```
{
"level":"error",
"ts":1628674622.422553,
"logger":"http.log.access.log1",
"msg":"handled request",
"request":{
	"remote_addr":"IP.DEL.MIO.CELLULARE:23697",
	"proto":"HTTP/2.0",
	"method":"GET",
	"host":"ha2.pippo.dominio.net",
	"uri":"/",
	"headers":{
		"Accept-Language":["it-it"],
		"Accept-Encoding":["gzip, deflate, br"],
		"User-Agent":["Mozilla/5.0 (iPhone; CPU iPhone OS 14_7_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.1.2 Mobile/15E148 Safari/604.1"],
		"Accept":["text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8"]
		},
	"tls":{
		"resumed":false,
		"version":772,
		"cipher_suite":4867,
		"proto":"h2",
		"proto_mutual":true,
		"server_name":"ha2.pippo.dominio.net"
		}
	},
	"common_log":"IP.DEL.MIO.CELLULARE - - [11/Aug/2021:09:37:02 +0000] \"GET / HTTP/2.0\" 502 0",
	"duration":0.001043926,
	"size":0,
	"status":502,
	"resp_headers":{
		"Server":["Caddy"],
		"Referrer-Policy":["same-origin"],
		"Strict-Transport-Security":["max-age=31536000; includeSubdomains"],
		"X-Content-Type-Options":["nosniff"],
		"X-Frame-Options":["SAMEORIGIN"],
		"X-Xss-Protection":["1; mode=block"]
	}
}
```
Questo e' il file di configurazione di Home Assistant **configuration.yaml**
```
http:
  server_port: 7000
  use_x_forwarded_for: true
  trusted_proxies: 
    - 172.17.0.0/16
    - 192.168.1.67
    - 127.0.0.1
```


Infine:

> Lato Caddy non vedo alcun servizio configurato sulla porta 7000

Ecco, allora non ho capito, credevo che  `reverse_proxy http:127.0.0.1:7000`  servisse a questo.

> se ti risponde HA vuol dire quindi che da Caddy nemmeno ci passi

Esatto.
A questo punto penso che ci sia qualche errore nella configurazione.

Questa e' la parte di docker-compose.yaml riguardo **duckdns** e **caddy**
```
duckdns:
    container_name: duckdns
    image: ghcr.io/linuxserver/duckdns
    environment:
    - "TZ=Europe/Rome"
    - "SUBDOMAINS=xxxxxxxxxxxxx"
    - "TOKEN=xxxxxxxxxxxxxxxxxxxxxxxxxx"
    restart: always

  caddy:
    container_name: caddy
    image: caddy:latest
    volumes:
    - "/home/pi/caddy/Caddyfile:/etc/caddy/Caddyfile"
    - "/home/pi/caddy/data:/data"
    environment:
    - "TZ=Europe/Rome"
    restart: always
    ports:
    - "80:80"
    - "443:443"
    - "9000:9000"
    - "9001:9001"
    - "9002:9002"
    - "9003:9003"
    - "9004:9004"
```

Ho riverificato anche i certificati e sembra tutto OK.
Ho anche provato a attivare DMZ per capire se ci fosse un problema nel Fastgate.

**Reply by snakuzzo** [reply]
Innanzi tutto ti chiedo la cortesia di non rispondere in nuovi thread ma di fare reply direttamente sull'ultimo messaggio


Poi...la configurazione del container con cui gestisci duckdns non mi dice nulla così. Devi fare in modo che il tuo IP pubblico sia associato al tuo (o ai tuoi) domini. Che sia con DuckDNS o con altri servizi o che sia con un container docker creato con un'immagine anziché un'altra questo non interessa...saprai tu come gestirla.

Per quanto riguarda Caddy ed HA ...tu avevi scritto...

`Se apro la porta 7000 sul router allora https://pippo.dominio.net:7000 si collega correttamente ad HA`

Se fai così ti torna che non passi da Caddy ma vai diretto su HA ? 

E se fai così ti tornerà che quasi sicuramente avrai configurato HA direttamente in https

E se così fosse, come cavolo farebbe Caddy ad inoltrare la richiesta verso la tua istanza di HA in http se il servizio sotto risponde in https ?

Ricontrolla tutte le configurazioni. Per fare questo giro devi configurare HA in HTTP e NON in HTTPS. L'HTTPS lo gestirà solo Caddy.

Consiglio da amico...prima di fare 40000 prove tutte insieme, fanne funzionare UNA...quella che ti serve

**Reply by 2Max2Pax** [reply]
OK.

> Se fai così ti torna che non passi da Caddy ma vai diretto su HA ?

Certamente.

> E se fai così ti tornerà che quasi sicuramente avrai configurato HA direttamente in https

... mi sono perso. Intendi quello che ho scritto come _internal url_ ed _external url_ ?



Sto facendo molte prove perche' ho iniziato ad usare linux da meno di una settimana installando su un vecchio catorcio Ubuntu desktop e poi sostituito con Ubuntu server. (Prima di passare ad HA su PI + Zigbee voglio fare un po di test anche con i componenti che mi arriveranno nei prossimi giorni).

Comunque hai ragione, un passo per volta.

Quindi la prima cosa che vorrei capire e perche' questo non funziona:
```
https://pippo.dominio.net:9000 {
  respond "Prova con porta"
  file_server
}
```

mentre questo funziona:
```
https://pippo.dominio.net {
  respond "Prova senza porta"
  file_server
}
```

In generale in qualsiasi prova fatta con `xxxxxxxxx:NUMERO PORTA` il browser mi restituisce
```
Impossibile raggiungere il sito
Connessione negata da xxxxxxxxxxx
```
E questo avviene sia usando duckdns che il mio indirizzo IP Pubblico. 
Come dicevo, se apro la porta 7000 per l'IP interno dove ho installato HA funziona sia con duckdns che con indirizzo ip pubblico.
Inoltre secondo [yougetsignal](https://www.yougetsignal.com/tools/open-ports/)  le porte 900x sono tutte chiuse.

Questo mi fa pensare di aver sbagliato qualcosa con Caddy, ma mi sono riletto tutte le procedure su [inDomus](https://indomus.it/guide/collegarsi-da-remoto-a-home-assistant-via-reverse-proxy/) e riverificato i certificati e mi sembra tutto OK.

**Reply by 2Max2Pax** [reply]
Al momento ho risolto la parte di reverse proxy verso HA.
Il problema era l'IP del container Caddy. Ecco come ho risolto:
`docker exec -it caddy ifconfig`
Questo comando ha restituito:
```
eth0      Link encap:Ethernet  HWaddr xx:xx:xx:xx:xx:xx
          inet addr:172.18.0.3  Bcast:172.18.255.255  Mask:255.255.0.0
```
L'IP del docker = 172.18.0.3
Quindi ho solamente dovuto modificare il file di configurazione di HA:
```
http:
  server_port: 7000
  use_x_forwarded_for: true
  trusted_proxies: 
    - 172.18.0.0/16
```

Rimane ancora da capire perche https://pippo.dominio.net:PORTA proprio non funziona.
Per ora sto utilizzando questo Caddyfile:
```
(https_header) {
  header {
    Strict-Transport-Security "max-age=31536000; includeSubdomains"
    X-XSS-Protection "1; mode=block"
    X-Content-Type-Options "nosniff"
    X-Frame-Options "SAMEORIGIN"
    Referrer-Policy "same-origin"
  }
}


https://ha.pippo.dominio.net { 
  import https_header
  reverse_proxy http://IP.DEL.SERVER.HA:7000
}

https://pippo.dominio.net { 
  respond "OK"
  file_server
}
```

**Reply by snakuzzo** [reply]
Ok ma, basandomi sul tuo ultimo post, non vedo alcuna configurazione https://pippo.dominio.net:PORTA

**Reply by 2Max2Pax** [reply]
Non l'ho copiata, deve fare solo un respond "OK 9000" su porta 9000.
P.S. ho provato anche con altre porte e sto facendo anche altri test, perche' non vorrei che il problema stia nel Fastgate.

**Reply by snakuzzo** [reply]
Copiato cosa? Non ho capito.
Ho solo detto che nell'ultimo blocco che hai postato, non vedo alcuna configurazione con la porta

**Reply by 2Max2Pax** [reply]
Ho copiato e postato solo la parte iniziale del Caddyfile.
Ho fatto anche un'altro test per capire se il problema sta nel Fastage che possa bloccare qualcosa, quindi ho usato la "_configurazione semplificata porte_" abilitando il servizio AIM talk che usa le porte 4099 e 5190, ma non ha funzionato.
Mi rimane solo il dubbio che il problema stia nel fatto che non posso fare forwarding della porta 80 perche il Fastagte non lo permette.

**Reply by snakuzzo** [reply]
però scusami, fai un po' d'ordine.
Ora la porta 80 da dove è uscita?
Se vuoi fare prove per conto tuo...fai pure, però poi non incasinare tutto qui :D

**Reply by 2Max2Pax** [reply]
Siccome non riesco a capire che cosa non funziona ho ipotizzato che potesse essere legato alla porta 80 (ma non credo sia quello il problema) .

E' normale questo risultato per il comando **caddy --validate** ?
```
[WARNING][caddyfile] Caddyfile:2: input is not formatted with 'caddy fmt'
INFO    http    enabling automatic HTTP->HTTPS redirects        {"server_name": "srv0"}
INFO    http    server is listening only on the HTTPS port but has no TLS connection policies; adding one to enable TLS {"server_name": "srv1", "https_port": 443}
```
Mi riferisco soprattutto all'ultima riga.

P.S. Caddy versione **v2.4.3**

**Reply by snakuzzo** [reply]
scusami...ma stai facendo redirect da HTTP ad HTTPS ?
Se è così devi fare come scritto nel messaggio di errore.

**Reply by 2Max2Pax** [reply]
> scusami...ma stai facendo redirect da HTTP ad HTTPS ?

Mi stai sopravvalutando ... e se lo stessi facendo, non saprei come...

> Se è così devi fare come scritto nel messaggio di errore.

Anche in questo caso ... non saprei che cosa fare. Ho trovato la documentazione su Caddy un po lacunosa, probabilmente presuppone che che l'utente abbia una conoscenza di base, che io non ho.

Comunque per ora getto la spugna.
In un prossimo futuro rifarò tutto da capo, installazione pulita, probabilmente su PI.

Adesso sto sperimentando, visto che devo decidere che tipologia di componentistica usare anche se la strada verso Zigbee sembra gia segnata per via della compatibilita' del mio impianto con la loro serie View Wireless e quindi gli interventi sull'impianto dovrebbero essere minimi (molti meno cavi da tirare). Mi e' anche arrivato uno Shelly 2.5 che voglio testare su una tapparella per impratichirmi con MQTT.

Comunque Diego, grazie per l'aiuto.
